js/src/admin/components/ExtensionPage.js

```diff
@@ -0,0 +1,384 @@
+      items.add(
+        field,
+        LinkButton.component(
+          {
+            href: value.href,
+            icon: value.icon,
+            external: true,
+            target: '_blank',
+          },
+          app.translator.trans(`core.admin.extension.info_links.${field}`)
+        )
+      );
+    });
+
+    return items;
+  }
+
+  submitButton() {
+    return (
+      <Button onclick={this.saveSettings.bind(this)} className="Button Button--primary" loading={this.loading} disabled={!this.changed()}>
+        {app.translator.trans('core.admin.settings.submit_button')}
+      </Button>
+    );
+  }
+
+  /**
+   * getSettings accepts an array of settings.
+   * Depending on the type of input, you can set the type to 'bool', 'select', or
+   * any standard <input> type.
+   *
+   * @example
+   * [
+   *   {
+   *      setting: 'acme.checkbox',
+   *      label: app.translator.trans('acme.admin.setting_label'),
+   *      type: 'bool'
+   *   }
+   * ]
+   *
+   * @example
+   * [
+   *   {
+   *      setting: 'acme.select',
+   *      label: app.translator.trans('acme.admin.setting_label'),
+   *      type: 'select',
+   *      options: {
+   *        'option1': 'Option 1 label',
+   *        'option2': 'Option 2 label',
+   *      },
+   *      default: 'option1',
+   *   }
+   * ]
+   *
+   * @param settings
+   * @returns {JSX.Element}
+   */
+  getSettings(settings) {
+    return settings.map((entry) => {
+      const setting = entry.setting;
+      const value = this.setting([setting])();
+      if (['bool', 'checkbox', 'switch', 'boolean'].includes(entry.type)) {
+        return (
+          <div className="Form-group">
+            <Switch state={!!value && value !== '0'} onchange={this.settings[setting]}>
+              {entry.label}
+            </Switch>
+          </div>
+        );
+      } else if (['select', 'dropdown', 'selectdropdown'].includes(entry.type)) {
+        return (
+          <div className="Form-group">
+            <label>{entry.label}</label>
+            <Select value={value || entry.default} options={entry.options} buttonClassName="Button" onchange={this.settings[setting]} />
+          </div>
+        );
+      } else {
+        return (
+          <div className="Form-group">
+            <label>{entry.label}</label>
+            <input type={entry.type} className="FormControl" bidi={this.setting(setting)} />
+          </div>
+        );
+      }
+    });
+  }
+
+  toggle() {
+    const enabled = this.isEnabled();
+
+    this.changingState = true;
+
+    app
+      .request({
+        url: app.forum.attribute('apiUrl') + '/extensions/' + this.extension.id,
+        method: 'PATCH',
+        body: { enabled: !enabled },
+        errorHandler: this.onerror.bind(this),
+      })
+      .then(() => {
+        if (!enabled) localStorage.setItem('enabledExtension', this.extension.id);
+        window.location.reload();
+      });
+
+    app.modal.show(LoadingModal);
+  }
+
+  dirty() {
```

Comment:
We have `isEnabled` below but the `dirty` and `changed` methods aren't prefixed by `is` - we should do one or the other, not both imo.

Reply:
You're right, `changed` should be `isChanged`, but `dirty` doesn't return a bool so it shouldn't be prefixed with is.

Reply:
Ah, good point.